### PR TITLE
added integrity protection patch

### PR DESCRIPTION
**Pull request by MareikeTaeubner**
added integrity protection patch

### CODE REVIEW

File: lib/browser/init.js [hunk 1]
```diff
@@ -103,14 +104,14 @@ require('./guest-window-manager')
 // Now we try to load app's package.json.
 let packagePath = null
 let packageJson = null
-const searchPaths = ['app', 'app.asar', 'default_app.asar']
+const searchPaths = asar.getSearchPath()
```
Comment (MaxFichtelmann) [inline lib/browser/init.js]: Why not just `['app.asar']`?

File: package.json [hunk 2]
```diff
@@ -37,6 +37,7 @@
   },
   "private": true,
   "scripts": {
+    "createPatchTar": "tar -cvzf patch.tgz $(git diff --diff-filter=ACMRT -r --name-only --no-commit-id v1.8.2 | xargs)",
```
Comment (MaxFichtelmann) [inline package.json]: a patch file would have been preferred IMO, anyway this script should not be commited

File: lib/browser/init.js [hunk 3]
```diff
@@ -163,7 +164,7 @@ require('./api/protocol')
 const mainStartupScript = packageJson.main || 'index.js'
 
 // Workaround for electron/electron#5050 and electron/electron#9046
-if (process.platform === 'linux' && ['Pantheon', 'Unity:Unity7', 'ubuntu:GNOME'].includes(process.env.XDG_CURRENT_DESKTOP)) {
+if (process.platform === 'linux' && ['Pantheon', 'Unity:Unity7', 'ubuntu:GNOME', 'pop:GNOME'].includes(process.env.XDG_CURRENT_DESKTOP)) {
```
Comment (MaxFichtelmann) [inline lib/browser/init.js]: is this really relevant to the integrity verification problem? which problem does this solve?

File: lib/common/asar.js [hunk 4]
```diff
@@ -1,20 +1,311 @@
...
+        // this listing.
+        // Also note, that we drop all other configuration property, if
+        // available (like `externalsAllowed`). Only "checksums" are used.
+        checksums: Object.assign(
+          {},
+          // additional files to check
+          // format: <filename>: <sha512 as base64 string>
+          // "configuration-manager.js": "not expected"
+          {
+            'identity-manager.yml': new Buffer(
```
Comment (MaxFichtelmann) [inline lib/common/asar.js]: are non-asar checksums needed? why this yml explicitely?

File: lib/common/asar.js [hunk 5]
```diff
@@ -1,20 +1,311 @@
...
+    // returns integrities as stringified json.
+    getIntegrity: function () {
+      if (typeof this.asarIntegrities === 'undefined') {
+        // default
+        this.asarIntegrities = {
+          checksums: {}
+        }
+
+        // get integrities from c/c++ source, if possible
+        const integrityRaw = asar.getIntegrity()
```
Comment (MaxFichtelmann) [inline lib/common/asar.js]: why is this needed and where does it come from?

File: lib/common/asar.js [hunk 6]
```diff
@@ -1,20 +1,311 @@
...
+          if (integrityConfig.externalAllowed === true) {
+            return;
+          }
+          throwErrorAndExit(
+            `${filePath} not located in the application resources path (${resourcesPath}) but loading of external files is disallowed (externalAllowed: ${
+              integrityConfig.externalAllowed
+            })`,
+            "EDAMAGEDAPP"
+          );
+        } */
```
Comment (MaxFichtelmann) [inline lib/common/asar.js]: please remove code that is commented out

File: lib/common/asar.js [hunk 7]
```diff
@@ -1,20 +1,311 @@
...
+
+        // find expected checksum by filename
+        const basename = path.basename(filePath)
+        const expected = integrityConfig.checksums[basename]
+
+        if (!expected) {
+          // NOTE in the renderer-process the "asar.getIntegrity" returns an empty list of checksums, because the bundle cannot be loaded.
+          // always fail for asar files (that's the minimum to be specified)
+          if (filePath.substr(-5) === '.asar') {
```
Comment (MaxFichtelmann) [inline lib/common/asar.js]: does this need to be restricted to asar file ending? is checking the filename enough?

File: lib/common/asar.js [hunk 8]
```diff
@@ -1,20 +1,311 @@
...
+        const actual = hashSync(fd)
+
+        // close file, if no FD was given
+        if (!fileFd) {
+          originalFs.closeSync(fd)
+        }
+
+        if (expected !== actual) {
+          // set checked to true for this file to avoid permanent checks with failures here!
+          checked[basename] = true
```
Comment (MaxFichtelmann) [inline lib/common/asar.js]: this seems fragile - allows subsequent reads to succeed

**Review by MaxFichtelmann — CHANGES_REQUESTED** [review]

File: lib/browser/init.js (hunk 1 above)
Comment (kloener) [reply lib/browser/init.js]: This file is not part of the binary (like "asar.js"). So you're probably able to modify the "electron.asar", where the "searchPaths" can be found in plaintext (if you do so).

Therefore, the "getSearchPaths" is required (see https://github.com/procilon/electron/pull/2/files#diff-30497f0d340d872cd5c3c0d6a87cfa0bR161). 

Proof, that "init.js" lies in electron.asar:
```bash
➜  Resources git:(master) pwd
/...some-electron-project.../dist/mac/App.app/Contents/Resources
➜  Resources git:(master) cat electron.asar | grep "const searchPaths = "
Binary file (standard input) matches
```

File: lib/browser/init.js (hunk 3 above)
Comment (kloener) [reply lib/browser/init.js]: It is not relevant for the integrity verification.

(my changes were previously made on Master-Branch of electron)

File: lib/common/asar.js (hunk 4 above)
Comment (kloener) [reply lib/common/asar.js]: The yml was a test file, if all loaded files (with fs.*) are recognized by the verification (yes, it is). But yes, non-asar checksums can be required by some projects. At least you've to provide the asar checksums, though.

File: lib/common/asar.js (hunk 5 above)
Comment (kloener) [reply lib/common/asar.js]: Currently, it isn't needed. Except for macOS applications: the electron-builder puts its own checksums to the "Info.plist", which are prioritized to the manually added ones.

File: lib/common/asar.js (hunk 7 above)
Comment (kloener) [reply lib/common/asar.js]: Yes, it has to be restricted to asar file ending. Otherwise, you have to create checksums for all files, even log or config files, which can be modified at runtime or by the user.

Checking the filename should be enough. "splitPath" does the same here: https://github.com/procilon/electron/blob/8b097e824d8ce82f686aebc956d5d69798325189/lib/common/asar.js#L347

File: lib/common/asar.js (hunk 8 above)
Comment (kloener) [reply lib/common/asar.js]: Should be enough security IMHO. If you restart the application all checks are made again. Changes are only passing, if made at runtime after a check succeeded. Additionally, the "app.asar" can get a huge size, which will be checked over and over again, if this line is removed.

You can enable the logs locally, if you wanna see the requests to app.asar: https://github.com/procilon/electron/blob/8b097e824d8ce82f686aebc956d5d69798325189/lib/common/asar.js#L306

**Comment by kloener** [comment]
I think some files are missing. See https://github.com/procilon/electron/pull/2/files?diff=unified#diff-16d4b6f742a19c8f163952914fd6cdc1R416.

* 'atom/common/asar/manifest_asar.h'
* 'atom/common/asar/manifest_asar_mac.mm'
* 'atom/common/asar/manifest_asar_win.cc'
* 'atom/common/asar/manifest_asar_linux.cc'

They ain't doing anything, except for the mac file, but they're required, I guess.

File: lib/browser/init.js (hunk 1 above)
Comment (MaxFichtelmann) [reply lib/browser/init.js]: we will investigate further. But as long as the code in electron.asar is loaded with integrity verification js code is just fine.
It feels to me as if changes to C++ code are not needed.

File: lib/common/asar.js [hunk 9]
```diff
@@ -1,15 +1,187 @@
...
+      // make asynchronous call
+      setTimeout(function () {
+        // Setting fs reference, if not already done
+        if (_originalFs && originalFs == null) {
+          originalFs = _originalFs
+        }
+
+        // integrityConfig will be loaded once
+        if (!integrityConfig) {
+          if (integrityConfig === null) {
```
Comment (kloener) [inline lib/common/asar.js]: This block can never be reached anymore.

**Review by MaxFichtelmann — APPROVED** [review]
